### ISSUE TITLE
[QUESTION]Loss Differences Between Single GPU and Tensor Parallelism

### ISSUE THREAD

**Issue by Tona8941** [NVIDIA/Megatron-LM#1736]
**Your question**
I want to understand if the loss differences I'm observing between single GPU and tensor parallel execution are reasonable/expected, and how to determine whether these are acceptable systematic variation or silent failures. Are there established ranges or detection methods to validate the loss accuracy of  parallelisms?

**Description**
I've observed consistent differences in loss values between single GPU and tensor parallel (TP=2) training, even in the most minimal model configurations. I created a systematic set of experiments to isolate the source of these differences and would like to understand if this is expected behavior and how to properly validate numerical accuracy in tensor parallel settings.

**Environment**

- Container Image: nvcr.io/nvidia/pytorch:23.12-py3
- PyTorch version: 2.2.0a0+81ea7a4 (NGC 23.12 container version)
- CUDA version: 12.3
- cuDNN version: 8.9.6
- NCCL version: 2.19.3
- Python version: 3.10
- Number of GPUs: 2

**Steps to reproduce**
The test scripts should be run from a directory at the same level as the megatron-lm folder:
workspace/
├── megatron-lm/
├── test_scripts.py
└── logs/

```
#Single GPU
#!/bin/bash
export CUDA_DEVICE_MAX_CONNECTIONS=1
export CUBLAS_WORKSPACE_CONFIG=:4096:8
export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:False

python ./megatron-lm/pretrain_gpt.py \
    --num-layers 1 \
    --hidden-size 64 \
    --num-attention-heads 2 \
    --seq-length 32 \
    --max-position-embeddings 32 \
    --vocab-size 128 \
    --micro-batch-size 1 \
    --global-batch-size 1 \
    --train-samples 20 \
    --lr 1e-6 \
    --seed 1234 \
    --attention-softmax-in-fp32 \
    --use-cpu-initialization \
    --attention-dropout 0.0 \
    --hidden-dropout 0.0 \
    --init-method-std 0.02 \
    --optimizer adam \
    --tokenizer-type GPT2BPETokenizer \
    --data-path ./data/test_data_text_document \
    --vocab-file ./vocab/gpt2-vocab.json \
    --merge-file ./vocab/gpt2-merges.txt \
    --save ./checkpoints/single \
    --log-interval 1 \
    --save-interval 100 \
    --no-load-optim \
    --no-load-rng 2>&1 | tee single_gpu.log

#Tensor Parallel
#!/bin/bash
export CUDA_DEVICE_MAX_CONNECTIONS=1
export CUBLAS_WORKSPACE_CONFIG=:4096:8
export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:False
export NCCL_ALGO=Tree

torchrun --nnodes=1 --nproc-per-node=2 ./megatron-lm/pretrain_gpt.py \
    --tensor-model-parallel-size 2 \
    --pipeline-model-parallel-size 1 \
    --num-layers 1 \
    --hidden-size 64 \
    --num-attention-heads 2 \
    --seq-length 32 \
    --max-position-embeddings 32 \
    --vocab-size 128 \
    --micro-batch-size 1 \
    --global-batch-size 1 \
    --train-samples 20 \
    --lr 1e-6 \
    --seed 1234 \
    --attention-softmax-in-fp32 \
    --use-cpu-initialization \
    --attention-dropout 0.0 \
    --hidden-dropout 0.0 \
    --init-method-std 0.02 \
    --optimizer adam \
    --tokenizer-type GPT2BPETokenizer \
    --data-path ./data/test_data_text_document \
    --vocab-file ./vocab/gpt2-vocab.json \
    --merge-file ./vocab/gpt2-merges.txt \
    --save ./checkpoints/tensor_parallel \
    --log-interval 1 \
    --save-interval 100 \
    --no-load-optim \
    --no-load-rng 2>&1 | tee tensor_parallel.log
```

**Results**
I conducted several controlled experiments to isolate the source of differences:
```
Experiment              Description                      Single GPU    TP=2          Observation
----------------------  -------------------------------  ------------  ------------  -------------------------
minimal_single_layer    Baseline 1-layer transformer    1.082085e+01  1.084218e+01  Consistent difference
linear_attention        Disabled softmax computation     1.082085e+01  1.084218e+01  Same difference as baseline
fixed_weights          Zero initialization              1.082584e+01  1.082838e+01  Reduced but present
attention_heads_test   Increased attention heads (2→4)  1.092446e+01  1.088210e+01  Larger absolute difference
```
**Analysis**
The differences appear to originate from the AllReduce operations during tensor parallel communication, as:

- They persist even without softmax operations
- They scale with the degree of parallelism
- They're affected by numerical magnitudes
- They occur even with deterministic settings (fixed seed, CPU initialization)

**Questions**

1. Are these differences originating from AllReduce communication operations?

- My experiments show the differences persist even when disabling softmax, suggesting they come from the communication layer rather than specific computational kernels.

2. Why are these numerical differences unavoidable in tensor parallelism, and what is their practical impact?

- Is this due to floating-point non-associativity in distributed reductions?
- Will these differences affect final model quality or training convergence?

3. How can we distinguish between acceptable numerical variance and silent errors?

- Are there tools or testing frameworks within Megatron-LM to validate numerical accuracy?
- What is the acceptable error range (e.g., relative difference threshold)?
- Is there a recommended test suite or validation procedure to ensure tensor parallel execution is working correctly versus experiencing silent failures?

**Comment by sbhavani** [comment]
@Tona8941 Thanks for the detailed report. The zero-initialization results are consistent with TP-dependent vocabulary padding, but given the age of the issue and missing exact Megatron-LM revision, we cannot confirm this as the complete root cause.

You could also try [Megatron Bridge](https://github.com/NVIDIA-NeMo/Megatron-Bridge), which provides explicit tokenizer and model vocabulary configuration.

Closing as stale. If this persists on current main, please open a new issue with the exact versions.
